metricflow/inference/solver/weighted_tree.py

```diff
@@ -0,0 +1,163 @@
+    InferenceSignalType,
+)
+from metricflow.inference.solver.base import InferenceSolver
+
+
+NodeWeighterFunction = Callable[[InferenceSignalConfidence], int]
+
+
+class WeightedTypeTreeInferenceSolver(InferenceSolver):
+    """Assigns weights to each type in the column type tree and attemptins to traverse it using a weight percentage threshold."""
+
+    @staticmethod
+    def default_weighter_function(confidence: InferenceSignalConfidence) -> int:
+        """The default weighter function.
+
+        It assigns weights 1, 2, 3 and 5 for LOW, MEDIUM, HIGH and FOR_SURE confidences, respectively. It then sums
+        the weights of all provided confidence scores.
+        """
+        confidence_weight_map = {
+            InferenceSignalConfidence.LOW: 1,
+            InferenceSignalConfidence.MEDIUM: 2,
+            InferenceSignalConfidence.HIGH: 4,
+            InferenceSignalConfidence.FOR_SURE: 8,
+        }
+
+        return confidence_weight_map[confidence]
+
+    def __init__(
+        self, weight_percent_threshold: float = 0.6, weighter_function: Optional[NodeWeighterFunction] = None
+    ) -> None:
+        """Initialize the solver.
+
+        weight_percent_threshold: a number between 0.5 and 1. If a node's weight corresponds to a percentage
+            above this threshold with respect to its siblings' total weight sum, the solver will progress deeper
+            into the type tree, entering that node. If not, it stops at the parent.
+        weighter_function: a function that returns a weight given a confidence score. It will be used
+            to assign integer weights to each node in the type tree based on its input signals.
+        """
+        assert weight_percent_threshold >= 0.5 and weight_percent_threshold <= 1
+        self._weight_percent_threshold = weight_percent_threshold
+
+        self._weighter_function = (
+            weighter_function
+            if weighter_function is not None
+            else WeightedTypeTreeInferenceSolver.default_weighter_function
+        )
+
+    def _get_cumulative_weights_for_root(
+        self,
+        root: InferenceSignalNode,
+        weights: Dict[InferenceSignalNode, int],
+        non_complimentary_weights: Dict[InferenceSignalNode, int],
+        signals_by_type: Dict[InferenceSignalNode, List[InferenceSignal]],
+    ) -> Dict[InferenceSignalNode, int]:
+        """Get a dict of cumulative weights, starting at `root`.
+
+        A parent node's weight is the sum of all its children plus its own weight. Complimentary children get
+        excluded from the parent's sum.
+
+        root: the root to start assigning cumulative weights from.
+        weights: the output dictionary to assign the cumulative weights to
+        non_complimentary_weights: similar to weights, but the weight of each node excludes the weights
+            of all of its complimentary children
+        signals_by_type: a dictionary that maps signal type nodes to signals
+        """
```

Comment:
I like to make the function signatures tell the callers what to expect even if they don't read the docstring, because most people won't - and here maybe you set input defaults to defaultdict(int) (or maybe not, depends on what you want from the typing):

```suggestion
    def _get_cumulative_weights_for_root(
        self,
        root: InferenceSignalNode,
        output_weights: Dict[InferenceSignalNode, int],
        output_parent_only_weights: Dict[InferenceSignalNode, int],
        signals_by_type: Dict[InferenceSignalNode, List[InferenceSignal]],
    ) -> Dict[InferenceSignalNode, int]:
        """Get a dict of cumulative weights, starting at `root`.

        A parent node's weight is the sum of all its children plus its own weight. Children tagged as only
        applying to their parent are excluded from their grand-parent's sum.

        root: the root to start assigning cumulative weights from.
        output_weights: the output dictionary to assign the cumulative weights to
        output_parent_only_weights: similar to output_weights, but the weight of each node excludes the weights
            of all of its parent-only children
        signals_by_type: a dictionary that maps signal type nodes to signals
        """
```